Snowflake.Data/Core/Session/ConnectionPoolManager.cs

```diff
@@ -0,0 +1,159 @@
+            var values = _pools.Values.Select(it => it.GetMaxPoolSize()).Distinct().ToList();
+            switch (values.Count)
+            {
+                case 0:
+                    return SFSessionHttpClientProperties.DefaultMaxPoolSize;
+                case 1:
+                    return values.First();
+                default:
+                    throw new SnowflakeDbException(SFError.INCONSISTENT_RESULT_ERROR, "Multiple pools have different Max Pool Size values");
+            }
+        }
+
+        public void SetTimeout(long connectionTimeout)
+        {
+            throw s_operationNotAvailable;
+        }
+
+        public long GetTimeout()
+        {
+            s_logger.Debug("ConnectionPoolManager::GetTimeout");
+            var values = _pools.Values.Select(it => it.GetTimeout()).Distinct().ToList();
+            switch (values.Count)
+            {
+                case 0:
+                    return (long) SFSessionHttpClientProperties.DefaultExpirationTimeout.TotalSeconds;
+                case 1:
+                    return values.First();
+                default:
+                    throw new SnowflakeDbException(SFError.INCONSISTENT_RESULT_ERROR, "Multiple pools have different Timeout values");
+            }
+        }
+
+        public int GetCurrentPoolSize()
+        {
+            s_logger.Debug("ConnectionPoolManager::GetCurrentPoolSize");
+            return _pools.Values.Select(it => it.GetCurrentPoolSize()).Sum();
+        }
+
+        public bool SetPooling(bool poolingEnabled)
+        {
+            throw s_operationNotAvailable;
+        }
+
+        public bool GetPooling()
+        {
+            s_logger.Debug("ConnectionPoolManager::GetPooling");
+            return true; // in new pool pooling is always enabled by default, disabling only by connection string parameter
+        }
+
+        public SessionPool GetPool(string connectionString, SecureString password)
+        {
+            s_logger.Debug($"ConnectionPoolManager::GetPool");
+            var poolKey = GetPoolKey(connectionString, password);
+
+            if (_pools.TryGetValue(poolKey, out var item))
+            {
+                item.ValidateSecurePassword(password);
+                return item;
+            }
+
+            lock (s_poolsLock)
+            {
+                if (_pools.TryGetValue(poolKey, out var poolCreatedWhileWaitingOnLock))
+                {
+                    poolCreatedWhileWaitingOnLock.ValidateSecurePassword(password);
+                    return poolCreatedWhileWaitingOnLock;
+                }
+                s_logger.Info($"Creating new pool");
+                var pool = SessionPool.CreateSessionPool(connectionString, password);
+                _pools.Add(poolKey, pool);
+                return pool;
+            }
+        }
+
+        public SessionPool GetPool(string connectionString)
+        {
+            s_logger.Debug($"ConnectionPoolManager::GetPool");
```

Comment:
maybe we could add to log "GetPool for connection string"?

Reply:
done